apps/mark-scan/backend/src/custom-paper-handler/state_machine.ts

```diff
@@ -1,267 +1,394 @@
-    await this.driver.setLineSpacing(0);
-    debug('set line spacing to 0');
-    await this.driver.setPrintingSpeed('slow');
-    debug('set printing speed to slow');
+  stopMachineService(): void {
+    this.machineService.stop();
   }
 
-  // Hacky "state machine"
-  async getSimpleStatus(): Promise<SimpleServerStatus> {
-    // detailed sensor state
-    const paperHandlerStatus = await this.driver.getPaperHandlerStatus();
+  // Leftover wrapper. Keeping this so the interface between API and state machine is the same until
+  // I can get around to migrating it later in the PR
+  getSimpleStatus(): SimpleServerStatus {
+    const { state } = this.machineService;
+    debug(`getSimpleStatus polled. state=${state.value}`);
 
-    switch (this.status) {
-      case 'no_paper':
-        if (isPaperReadyToLoad(paperHandlerStatus)) {
-          this.status = 'paper_ready_to_load';
-        }
-        break;
-      case 'paper_ready_to_load':
-        if (!isPaperReadyToLoad(paperHandlerStatus)) {
-          this.status = 'no_paper';
-        }
-        break;
-      case 'parking_paper':
-        if (paperHandlerStatus.parkSensor) {
-          this.status = 'paper_parked';
-        } else if (Date.now() - this.lastActionInitiatedTime > 10_000) {
-          this.status = 'no_paper';
-        }
-        break;
-      case 'paper_parked':
-        // Nothing to do - frontend triggers status change. This will change soon so not
-        // worth centralizing machine logic
-        break;
-      case 'printing_ballot':
-        if (!isPaperInScanner(paperHandlerStatus)) {
-          this.status = 'ballot_printed';
-        }
-        break;
-      case 'ejecting':
-        if (!isPaperAnywhere(paperHandlerStatus)) {
-          this.status = 'no_paper';
-        }
-        break;
-      case 'ballot_printed':
-        // Unimplemented
-        break;
+    switch (true) {
+      case state.matches('no_paper'):
+        return 'no_paper';
+      case state.matches('loading_paper'):
+        return 'loading_paper';
+      case state.matches('waiting_for_ballot_data'):
+        return 'waiting_for_ballot_data';
+      case state.matches('printing_ballot'):
+        return 'printing_ballot';
+      case state.matches('scanning'):
+        return 'scanning';
+      case state.matches('interpreting'):
+        return 'interpreting';
+      case state.matches('presenting_ballot'):
+        return 'presenting_ballot';
       default:
-        throwIllegalValue(this.status);
+        return 'no_hardware';
     }
-    return this.status;
   }
 
-  resetStatus(): void {
-    this.status = 'no_paper';
-  }
+  printBallot(
+    pdfData: Uint8Array,
+    options: Partial<ImageConversionOptions> = {}
+  ): Promise<void> {
+    this.machineService.send({
+      type: 'VOTER_INITIATED_PRINT',
+    });
 
-  async logStatus(): Promise<void> {
-    debug('%O', await this.driver.getPaperHandlerStatus());
+    return driverPrintBallot(this.driver, pdfData, options);
```

Comment:
Should this printing method get called from within the state machine?

The approach we used in apps/scan was to have the state machine manage all of the business logic, including triggering side effects, so that it could all be understood in one place.

Could that approach be applied here?

Reply:
Do you mean just the `driverPrintBallot` call? We could definitely move that call to a state; I agree it's cleaner that way.

Not sure if the way I support outside interaction with the state machine here is the best way. The goal is to expose a method to allow the API to cause state transitions (in this case from WaitingForBallotData to Printing). Just wrapping `send` seemed like the most straightforward way but if there's a better way let me know.

Reply:
Yes, I meant the `driverPrintBallot` call.

Wrapping `send` seems like the right approach for managing outside interaction. Pretty sure that's how I did it in `apps/scan`.